handlers/photo-audio-haptics-handler/src/server.ts

```diff
@@ -0,0 +1,267 @@
+    const objGeometryData: utils.objGeometryInfo[] = [];
+    ttsData.push({ "value": utils.generateIntro(preSecondCat), "type": "text" });
+    if (preSemSeg) {
+        // Use all segments returned for now.
+        // Filtering may be helpful later.
+        const [ttsInfo, geometryInfo] = utils.generateSemSeg(preSemSeg);
+        ttsData.push(...ttsInfo);
+        segGeometryData.push(...geometryInfo);
+
+        if (preObjDet && preGroupData) {
+            ttsData.push({ "value": "It also", "type": "text" });
+        }
+    }
+    if (preObjDet && preGroupData) {
+        const [ttsInfo, geometryInfo] = utils.generateObjDet(preObjDet, preGroupData);
+        ttsData.push(...ttsInfo);
+        objGeometryData.push(...geometryInfo);
+    }
+
+    // Concatenate adjacent text entries
+    for (let i = 0; i < ttsData.length - 1; i++) {
+        if (ttsData[i].type === "text" && ttsData[i + 1].type === "text") {
+            ttsData[i].value += " " + ttsData[i + 1].value;
+            ttsData.splice(i + 1, 1);
+        }
+    }
+
+    // Generate rendering title
+    const renderingTitle = utils.renderingTitle(preSemSeg, preObjDet, preGroupData);
+
+    if (hasAudioHaptic) {
+        try {
+            // Do TTS
+            const ttsResponse = await utils.getTTS(ttsData.map(x => x["value"]));
+            // Add offset values to data
+            for (let i = 0, offset = 0; i < ttsData.length; i++) {
+                ttsData[i]["audio"] = {
+                    "offset": offset,
+                    "duration": ttsResponse.durations[i]
+                };
+                offset += ttsResponse.durations[i];
+            }
+
+            const scData = {
+                "data": ttsData,
+                "ttsFileName": ""
+            };
+
+            // Write to file
+            let inFile: string, outFile: string, jsonFile: string;
+            await fetch(ttsResponse["audio"]).then(resp => {
+                return resp.arrayBuffer();
+            }).then(async (buf) => {
+                inFile = filePrefix + req.body["request_uuid"] + ".wav";
+                await fs.writeFile(inFile, Buffer.from(buf));
+                scData["ttsFileName"] = inFile;
+                jsonFile = filePrefix + req.body["request_uuid"] + ".json";
+                await fs.writeFile(jsonFile, JSON.stringify(scData));
+                outFile = filePrefix + uuidv4() + ".flac";
+                await fs.writeFile(outFile, "");
+                await fs.chmod(outFile, 0o664);
+
+                console.log("Forming OSC...");
+                return utils.sendOSC(jsonFile, outFile, "supercollider", scPort);
+            }).then(async (entities: any) => {
+                const buffer = await fs.readFile(outFile);
+                // TODO detect mime type from file
+                const dataURL = "data:audio/flac;base64," + buffer.toString("base64");
+                if (hasAudioHaptic && entities.length > 0

```

Comment:
`entities.length > 0` seems redundant given the `entities.length > segGeometryData.length` check.